README.md
# ngx-infinite-scroller

Infinite and bidirectional scroll directive for Angular 11

All notable changes to this project are documented in [CHANGELOG.md](https://github.com/staskolukasz/ngx-infinite-scroller/blob/master/CHANGELOG.md) file.

## Installation

Run `npm install ngx-infinite-scroller --save` to install the library.

## Usage

### *.module.ts configuration

Include `NgxInfiniteScrollerModule` in your module

```typescript
import { NgxInfiniteScrollerModule } from 'ngx-infinite-scroller';

@NgModule({
  declarations: [],
  imports: [
    NgxInfiniteScrollerModule,
  ],
  providers: [],
  bootstrap: []
})
```

### *.component.html configuration

Include `ngxInfiniteScroller` directive in your `*.component.html` file

```html
<ul id="scroller"
    ngxInfiniteScroller
    strategy="scrollingToBoth"
    (onScrollUp)="onScrollUp()"
    (onScrollDown)="onScrollDown()">
  <li class="news"
      *ngFor="let item of news">
    {{item.title}}
  </li>
</ul>
```
By default the directive works as an infinite scroll from the top to the bottom of your list. To switch to other modes, use input parameters like:  
`strategy="scrollingToTop"`  
`strategy="scrollingToBottom"` (default)  
`strategy="scrollingToBoth"`  

Default value of initial scroll position depends on the strategy value:  
`strategy="scrollingToTop"` -> `initialScrollPosition="BOTTOM"`  
`strategy="scrollingToBoth"` -> `initialScrollPosition="MIDDLE"`  
`strategy="scrollingToBottom"` -> `initialScrollPosition="TOP"`  

It is also possible to customize the default behaviour using an optional input field:  
`initialScrollPosition="TOP"`  
`initialScrollPosition="MIDDLE"`  
`initialScrollPosition="BOTTOM"`  
`initialScrollPosition="50"` (position in px)  

### *.component.ts configuration

Handle `onScrollUp` and `onScrollDown` actions in your `*.component.ts` file

```typescript
import { Component, OnInit } from '@angular/core';
import { HttpClient } from '@angular/common/http';

import { Observable } from 'rxjs';
import { share, finalize } from 'rxjs/operators';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit {

  public news: Array<any> = [];

  private currentPage = 1;

  private request$: Observable<any>;

  constructor(private http: HttpClient) { }

  public ngOnInit() {
    this.getNews(this.currentPage)
      .pipe(finalize(() => this.onFinalize()))
      .subscribe((news) => {
        this.currentPage++;
        this.news = this.news.concat(news);
      });
  }

  public onScrollUp(): void {
    this.getNews(this.currentPage)
      .pipe(finalize(() => this.onFinalize()))
      .subscribe((news) => {
        this.currentPage++;
        this.news = news.concat(this.news);
      });
  }

  public onScrollDown(): void {
    this.getNews(this.currentPage)
      .pipe(finalize(() => this.onFinalize()))
      .subscribe((news) => {
        this.currentPage++;
        this.news = this.news.concat(news);
      });
  }

  // Prevent duplicate requests on scroll.
  // More: https://stackoverflow.com/a/50865911/6441494
  private getNews(page: number = 1): Observable<any> {
    if (this.request$) {
      return this.request$;
    } else {
      this.request$ = this.http.get(`https://node-hnapi.herokuapp.com/news?page=${page}`).pipe(share());
      return this.request$;
    }
  }

  private onFinalize(): void {
    this.request$ = null;
  }
}

```

### *.component.scss configuration

Add some styling in your `*.component.scss` file

```scss
#scroller {
  height: 100vh;
  width: 700px;
  overflow: scroll;
  padding: 0;
  margin: 0;
  list-style: none;
}

.news {
  padding: 30px;
}
```

## Development environment

Run `ng serve` for a dev server. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.

Run `npm run packagr` to build the library. The build artifacts will be stored in the `dist` directory.

## Publishing

Run `cd dist && npm publish` command to publish the package.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)

## [0.6.3] - 2020-11-25
### Changed
- Documentation fixes

## [0.6.2] - 2020-11-25
### Changed
- Update to Angular 11

## [0.6.1] - 2020-03-11
### Changed
- Documentation fixes

## [0.6.0] - 2020-03-11
### Changed
- Update to Angular 9

## [0.5.0] - 2019-09-23
### Changed
- Update to Angular 8
- Update documentation and sample application to prevent duplicate http requests on scroll event
- In server side rendering mutation observer is not used anymore.

## [0.4.5] - 2019-07-16
### Changed
- Fix npm vulnerables (lodash from 4.17.11 to 4.17.14)

## [0.4.4] - 2019-05-15
### Changed
- Fix npm vulnerables and npm update packages

## [0.4.1] - 2019-01-22
### Changed
- Readme.md updated (fix Angular version)

## [0.4.0] - 2019-01-22
### Added
- Added support of custom values of initial scroll position
- Readme.md updated (initialScrollPosition input)

### Changed
- Updated to Angular 7

## [0.3.5] - 2018-08-24
### Changed
- Readme.md updated (new rxjs syntax)
- NPM packages updated

## [0.3.4] - 2018-05-19
### Changed
- Changelog updated

## [0.3.0] - 2018-05-19
### Changed
- Updated to Angular 6

## [0.2.2] - 2018-03-12
### Changed
- BrowserModule removed from NgxInfiniteScrollerModule

## [0.2.1] - 2018-03-10
### Added
- CHANGELOG.md file
- DirectiveStateService (state management)
- ScrollHeightListener (used to listen to scrollbar height changes, instead of MutationObserver)

### Changed
- The way how previous scroll position is set
- Directive refactoring

src/example-host/app.component.ts
import { Component, OnInit } from '@angular/core';
import { HttpClient } from '@angular/common/http';

import { Observable } from 'rxjs';
import { share, finalize } from 'rxjs/operators';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit {

  public news: Array<any> = [];

  private currentPage = 1;

  private request$: Observable<any>;

  constructor(private http: HttpClient) { }

  public ngOnInit() {
    this.getNews(this.currentPage)
      .pipe(finalize(() => this.onFinalize()))
      .subscribe((news) => {
        this.currentPage++;
        this.news = this.news.concat(news);
      });
  }

  public onScrollUp(): void {
    this.getNews(this.currentPage)
      .pipe(finalize(() => this.onFinalize()))
      .subscribe((news) => {
        this.currentPage++;
        this.news = news.concat(this.news);
      });
  }

  public onScrollDown(): void {
    this.getNews(this.currentPage)
      .pipe(finalize(() => this.onFinalize()))
      .subscribe((news) => {
        this.currentPage++;
        this.news = this.news.concat(news);
      });
  }

  // Prevent duplicate requests on scroll.
  // More: https://stackoverflow.com/a/50865911/6441494
  private getNews(page: number = 1): Observable<any> {
    if (this.request$) {
      return this.request$;
    } else {
      console.log('Strona:', page);
      this.request$ = this.http.get(`https://node-hnapi.herokuapp.com/news?page=${page}`).pipe(share());
      return this.request$;
    }
  }

  private onFinalize(): void {
    this.request$ = null;
  }
}
